program_analysis/src/utils.ml

```diff
@@ -73,3 +73,213 @@ and pp_res fmt = function
+        match p with
+        | LabelResAtom _ | ExprResAtom _ ->
+            remove_edge pid rid;
+            let label =
+              Hashtbl.find edge_label (Format.sprintf "%s_%s" pid rid)
+            in
+            List.iter r ~f:(fun a ->
+                let aid = ida a in
+                add_edge pid aid;
+                (match label with
+                | Some label ->
+                    (* Format.printf "pid: %s | aid: %s | rid: %s\n" pid aid rid;
+                       Format.printf "edges labels:\n";
+                       Hashtbl.iteri edge_label ~f:(fun ~key ~data ->
+                           Format.printf "%s: %s\n" key data); *)
+                    ignore
+                      (Hashtbl.add edge_label
+                         ~key:(Format.sprintf "%s_%s" pid aid)
+                         ~data:label)
+                | None -> ());
+                dot_of_atom a new_p)
+        | _ ->
+            if List.length r = 1 then (
+              let a = List.hd_exn r in
+              let aid = ida a in
+              remove_edge pid rid;
+              add_edge pid aid;
+              (match
+                 Hashtbl.find edge_label (Format.sprintf "%s_%s" pid rid)
+               with
+              | Some label ->
+                  (* Format.printf "pid: %s | aid: %s | rid: %s\n" pid aid rid;
+                     Format.printf "edges labels:\n";
+                     Hashtbl.iteri edge_label ~f:(fun ~key ~data ->
+                         Format.printf "%s: %s\n" key data); *)
+                  ignore
+                    (Hashtbl.add edge_label
+                       ~key:(Format.sprintf "%s_%s" pid aid)
+                       ~data:label)
+              | None -> ());
+              (* needs to be called last to remove edges to PathCondAtoms *)
+              dot_of_atom a new_p)
+            else
+              List.iter r ~f:(fun a ->
+                  dot_of_atom a new_p;
+                  add_node (Format.sprintf "%s [label=\"|\"];" rid);
+                  add_edge rid (ida a)))
+  in
+  dot_of_res r None;
+  let nodes = Hashset.fold (Format.sprintf "%s\n%s") node_set "" in
+  let edges =
+    Hashtbl.fold edge_map ~init:"" ~f:(fun ~key ~data acc ->
+        String.Set.fold data ~init:acc ~f:(fun acc n ->
+            let edge_id = Format.sprintf "%s_%s" key n in
+            let label =
+              match Hashtbl.find edge_label edge_id with
+              | None -> ""
+              | Some l ->
+                  (* Format.printf "yo\n"; *)
+                  l
+            in
+            Format.sprintf "%s\n%s -> %s [label=\"%s\", fontsize=\"8\"]" acc key
+              n label))
+  in
+  let dot_file = Format.sprintf "graph%d.dot" test_num in
+  Out_channel.write_all dot_file
+    ~data:
+      (Format.sprintf
+         "digraph G {\n\
+          node [fontname=\"Arial\"]\n\
+          edge [fontname=\"Arial\"]\n\
+          %s%s\n\
+          }"
+         nodes edges);
```

Comment:
Generate a Dot file in the `_build/default/program_analysis/tests` directory, after this function (`dot_of_result`) is invoked through `dune test program_analysis`.